server/src/secret_exchange.rs

```diff
@@ -0,0 +1,131 @@
+// on a HashMap.
+// Before any transit operations the payload is base64 encoded and parsed into a
+// String.
+
+use std::collections::HashMap;
+
+use base64::prelude::*;
+use oo7::{crypto, Key};
+use zeroize::Zeroizing;
+
+const SECRET: &str = "secret";
+const PUBLIC: &str = "public";
+const PRIVATE: &str = "private";
+const IV: &str = "iv";
+const PROTOCOL: &str = "[sx-aes-1]\n";
+const CIPHER_TEXT_LEN: usize = 16;
+
+#[derive(Debug)]
+pub struct SecretExchange {
+    private_key: Key,
+    public_key: Key,
+}
+
+impl SecretExchange {
+    // Creates the initial payload containing caller public_key
+    pub fn begin(&self) -> String {
+        let map = HashMap::from([(PUBLIC, self.public_key.as_ref())]);
+
+        encode(&map)
+    }
+
+    // Creates the shared secret: an AES key
+    pub fn create_shared_secret(&self, exchange: &str) -> String {
+        let decoded = decode(exchange).unwrap();
+        let server_public_key = Key::new(decoded.get(PUBLIC).unwrap().to_vec());
+        let aes_key =
+            Key::generate_aes_key_for_secret_exchange(&self.private_key, &server_public_key);
+        let map = HashMap::from([(PRIVATE, aes_key.as_ref())]);
+
+        encode(&map)
+    }
+
+    pub fn new() -> Self {
+        let private_key = Key::generate_private_key();
+        let public_key = Key::generate_public_key_for_secret_exchange(&private_key);
+
+        Self {
+            private_key,
+            public_key,
+        }
+    }
+}
+
+// Converts a HashMap into a payload String
+fn encode(map: &HashMap<&str, &[u8]>) -> String {
+    let mut exchange = map
+        .iter()
+        .map(|(key, value)| format!("{}={}", key, BASE64_STANDARD.encode(value)))
+        .collect::<Vec<_>>()
+        .join("\n");
+    exchange.insert_str(0, PROTOCOL); // to add PROTOCOL prefix
+
+    exchange
+}
+
+// Converts a payload String into a HashMap
+fn decode(exchange: &str) -> Result<HashMap<&str, Vec<u8>>, base64::DecodeError> {
+    let (_, exchange) = exchange.split_once(PROTOCOL).unwrap(); // to remove PROTOCOL prefix
+    let pairs = exchange.split("\n").collect::<Vec<_>>();
+    let mut map: HashMap<&str, Vec<u8>> = HashMap::new();
+
+    for pair in pairs {
+        if pair.is_empty() {
+            // to avoid splitting an empty line (last new line)
+            break;
+        }
+        let (key, value) = pair.split_once("=").unwrap();
+        let encoded = BASE64_STANDARD.decode(value)?;
+        map.insert(key, encoded);
+    }
+
+    Ok(map)
+}
+
+// Retrieves the secret from final secret exchange string
+pub(crate) fn retrieve_secret(exchange: &str, aes_key: &str) -> Option<Zeroizing<Vec<u8>>> {
+    let decoded = decode(exchange).unwrap();
+    let secret = decoded.get(SECRET);
+
+    // If we cancel an ongoing prompt call, the final exchange won't have the secret
+    // or IV. The following is to avoid `Option::unwrap()` on a `None` value
+    secret?;
+
+    let secret = secret.unwrap();
```

Comment:
```suggestion
    // If we cancel an ongoing prompt call, the final exchange won't have the secret
    // or IV. The following is to avoid `Option::unwrap()` on a `None` value
    let secret = decoded.get(SECRET)?;

```